### ISSUE TITLE
About GRPO

### ISSUE THREAD

**Issue by BeerTai** [OpenRLHF/OpenRLHF#727]
Regarding the implementation of grpo, is it enough to just add the following code to the function `process_experiences()` in the current version? 
```
    def process_experiences(self, experiences: List[Experience]) -> Tuple[List[Experience], List[torch.Tensor]]:
        """
        Process experiences, this can be used to filter out some experiences or do some processing on the rewards.

        Output:
        - experiences: List of Experience
        - rewards: List of rewards
        """
        args = self.strategy.args
        # reward shaping for RLOO
        if args.advantage_estimator == "rloo":
            rewards = torch.cat([experience.info["reward"] for experience in experiences])
            rewards = rewards.reshape(-1, args.n_samples_per_prompt).to(device="cuda")
            baseline = (rewards.sum(-1, keepdim=True) - rewards) / (args.n_samples_per_prompt - 1)
            rewards = rewards - baseline
            rewards = rewards.flatten().to(device="cpu").chunk(len(experiences))
            return experiences, rewards

        if args.advantage_estimator == "group_norm":
            rewards = torch.cat([experience.info["reward"] for experience in experiences])
            rewards = rewards.reshape(-1, args.n_samples_per_prompt).to(device="cuda")
            rewards = (rewards - rewards.mean(1, keepdim=True)) / (rewards.std(1, keepdim=True) + 1e-8)
            rewards = rewards.reshape(-1).to(device="cpu").chunk(len(experiences))
            return experiences, rewards
        # default rewards
        return experiences, [experience.info["reward"] for experience in experiences]
```

**Comment by hijkzzz** [comment]
As long as you don't want to use K3 KL loss, this is how it is.

**Comment by wangzhen0518** [comment]
Is the following modification necessary and correct?

https://github.com/OpenRLHF/OpenRLHF/blob/cca506019f45e79c2a3df87494266552f47f43f9/openrlhf/trainer/ppo_utils/experience_maker.py#L229-L235

->

```python
elif self.advantage_estimator in ["reinforce", "rloo", "group_norm"]:
    ...
```

**Comment by hijkzzz** [comment]
Also see https://github.com/OpenRLHF/OpenRLHF/pull/730

**Comment by zzzichen277** [comment]
Will GRPO be supported in future updates? 
将来更新会支持grpo吗？

**Comment by Co1lin** [comment]
Hi @hijkzzz , I am wondering is the GRPO curve in the figure you shared above comes from "the original version", i.e. with `/ std` and K3 KL?

**Comment by hijkzzz** [comment]
GRPO merged https://github.com/OpenRLHF/OpenRLHF/commit/2c2aca7805cec89adf82d40acfe8c19d338e8613

**Comment by hijkzzz** [comment]
GRPO vs REINFORCE++

https://arxiv.org/abs/2502.14768

![Image](https://github.com/user-attachments/assets/76f01e3a-a0f8-4941-b603-21d5e3e2521d)


https://arxiv.org/abs/2502.01456

![Image](https://github.com/user-attachments/assets/a5325a3c-b657-4a3e-9f74-a564b2037fbd)
